### PR TITLE
CFE-3023: Retry SSL_read on SSL_ERROR_WANT_READ

### PR DESCRIPTION
**Pull request by nickanderson**
Changelog: Title

When the client gets SSL_ERROR_WANT_READ, we should re-try instead of shutting
down the whole session.

### CODE REVIEW

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2198/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2198/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2198/

File: libcfnet/tls_generic.c [hunk 1]
```diff
@@ -724,7 +724,20 @@ int TLSRecv(SSL *ssl, char *buffer, int toget)
...
     assert_SSLIsBlocking(ssl);
 
-    int received = SSL_read(ssl, buffer, toget);
+    int received,tries = 0,code=0;
+    do
+      {
+        received = SSL_read(ssl, buffer, toget);
+        if (received < 0)
+          {
+            code = TLSLogError(ssl, LOG_LEVEL_ERR, "SSL_read", received);
```
Comment (craigcomstock) [inline libcfnet/tls_generic.c]: Seems like LOG_LEVEL_ERR on each try could get rather spammy.

File: libcfnet/tls_generic.c [hunk 2]
```diff
@@ -724,7 +724,20 @@ int TLSRecv(SSL *ssl, char *buffer, int toget)
     assert(toget < CF_BUFSIZE);
     assert_SSLIsBlocking(ssl);
 
-    int received = SSL_read(ssl, buffer, toget);
+    int received,tries = 0,code=0;
+    do
+      {
+        received = SSL_read(ssl, buffer, toget);
```
Comment (craigcomstock) [inline libcfnet/tls_generic.c]: would SSL_peek() be a better choice here?

File: libcfnet/tls_generic.c [hunk 3]
```diff
@@ -724,7 +724,20 @@ int TLSRecv(SSL *ssl, char *buffer, int toget)
...
+        if (received < 0)
+          {
+            code = TLSLogError(ssl, LOG_LEVEL_ERR, "SSL_read", received);
+            if (code == SSL_ERROR_WANT_READ)
+              {
+                sleep(1);
+              }
+          }
+      } while( received < 0 && code == SSL_ERROR_WANT_READ && (tries++ < 10) );
+
```
Comment (craigcomstock) [inline libcfnet/tls_generic.c]: Would it be good to have the number 10 configurable? Also if still failed after 10 tries THEN log an error which states how many tries failed?

File: libcfnet/tls_generic.c [hunk 4]
```diff
@@ -724,7 +724,20 @@ int TLSRecv(SSL *ssl, char *buffer, int toget)
     assert(toget < CF_BUFSIZE);
     assert_SSLIsBlocking(ssl);
 
-    int received = SSL_read(ssl, buffer, toget);
+    int received,tries = 0,code=0;
+    do
```
Comment (craigcomstock) [inline libcfnet/tls_generic.c]: Why not use SSL_MODE_AUTO_RETRY? Or maybe this is another failure case you are working on?

File: libcfnet/tls_generic.c (hunk 2 above)
Comment (vpodzime) [reply libcfnet/tls_generic.c]: I think we can do `SSL_read()` just fine.

File: libcfnet/tls_generic.c [hunk 5]
```diff
@@ -724,7 +724,20 @@ int TLSRecv(SSL *ssl, char *buffer, int toget)
...
+      {
+        received = SSL_read(ssl, buffer, toget);
+        if (received < 0)
+          {
+            code = TLSLogError(ssl, LOG_LEVEL_ERR, "SSL_read", received);
+            if (code == SSL_ERROR_WANT_READ)
```
Comment (vpodzime) [inline libcfnet/tls_generic.c]: According to man:SSL_read(3), `SSL_Read()` can also return `SSL_ERROR_WANT_WRITE` so we need to check for that too.

File: libcfnet/tls_generic.c (hunk 4 above)
Comment (vpodzime) [reply libcfnet/tls_generic.c]: I think it's better to have the control over the retries.

File: libcfnet/tls_generic.c (hunk 3 above)
Comment (vpodzime) [reply libcfnet/tls_generic.c]: Yes, exactly.

**Comment by nickanderson** [comment]
> Also please try to follow the code style we are using.

maybe you can help me configure my emacs to do that automagically, I just selected it all and tabbed to indent :)

**Comment by vpodzime** [comment]
> maybe you can help me configure my emacs to do that automagically, I just selected it all and tabbed to indent :)

We have `contrib/cfengine-code-style.el`, but it doesn't take care of all the stuff, AFAICT.

**Comment by nickanderson** [comment]
Someone want to take over this PR? Maintainers can push to my branch or just branch from it

**Comment by vpodzime** [comment]
I can take this over and hopefully get it into a merge`able state next week.

**Comment by vpodzime** [comment]
Superseded by #3618

**Comment by cristiancardosodexcom** [comment]
I'm getting 

verbose: Copying remote file '192.168.1.1:masterfiles/cfe_internal/enterprise/federation/federation.cf', expecting 44872 bytes
 verbose: SSL read failed: receive timeout
 verbose: SSL read failed: receive timeout
 **verbose: SSL read failed: receive timeout**
 
 Any ideas guys?

**Comment by vpodzime** [comment]
> Any ideas guys?

Anything extraordinary in the logs on the hub? And is this not just bad network? Did you try `ping 192.168.1.1` and/or some other basic network tests?

**Comment by olehermanse** [comment]
@cristiancardosodexcom This old PR is probably not the best place for visibility, feel free to open a thread in our GitHub Discussions; https://github.com/cfengine/core/discussions

**Comment by cristiancardosodexcom** [comment]
@vpodzime yeah, ping responses, port open, firewall-cmd disabled, etc.

Thanks guys, yes I'll open a new discussion.

**Pull request closed without merging**